rlog/2024-06-24-gsub-largemsg-improvements.mdx

```diff
@@ -0,0 +1,396 @@
+where $F$ represents the number of peers included in floodpublish. 
+
+This arrangement works fine for relatively small/moderate message sizes. 
+However, as message sizes increase, a significant rise and fluctuations in network-wide message dissemination time are seen. 
+Interestingly, a higher $D$ or $F$ can also degrade performance in this situation. 
+
+Several aspects contribute to this behavior: 
+
+1. Ideally, a message transmission to a single peer concludes in $\tau_1 = \frac {L}{R}+P$ (ignoring any message processing time), 
+where $L, R, P$ represent message size, data rate, and link latency. 
+Therefore, the time required for sending a message on a 100Mbps link with 100ms latency 
+jumps from $\tau_1^{10KB} = 100.8ms$ for a 10KB message to $\tau_1^{1MB} = 180ms$ for a 1MB message. 
+For $D$ peers, the transmission time increases to $\tau_D^{1MB} = (80 \times D) + 100ms$, 
+triggering additional queuing delays (proportional to the transmission queue size) during each transmission round.  
+
+2. In practice, $\tau_1^{1MB}$ sometimes rises to several hundred milliseconds, 
+further exaggerating the abovementioned queuing delays.
+This rise is because TCP congestion avoidance limits the maximum in-flight bytes to approximately ${C_{wnd} \times MSS}$ in a single round trip time (RTT), 
+with $C_{wnd}$ increasing with the data transfer for each flow. 
+Consequently, sending the same message through a newly established (cold) connection takes longer. 
+The message transfer time lowers as the $C_{wnd}$ grows. 
+Therefore, performance-improvement practices such as floodpublish, frequent mesh adjustment, and lazy sending 
+typically result in longer than expected message dissemination times for large messages (due to cold connections). 
+It is also worth mentioning that some TCP variants reset their $C_{wnd}$ after different periods of inactivity.
+
+3. Theoretically, the message transmission time to D peers $(\tau_D)$ remains the same, 
+even if the message is relayed sequentially to all peers or a simultaneous transmission is carried out.
+However, sequential transmissions finish early for individual peers, allowing them to relay early.
+It may result in quicker network-wide message dissemination. 
+
+4. A realistic network comprises nodes with dissimilar capabilities (bandwidth, link latency, compute, etc.).
+As the message disseminates, it's not uncommon for some peers to receive it much earlier than others. 
+Early gossip (IHAVE announcements) may bring in many IWANT requests to the early receivers (even from peers already receiving the same message), 
+which adds to their workload.  
+
+5. A busy peer (with a sizeable outgoing message queue) will enqueue (or simultaneously transfer) newly scheduled outgoing messages. 
+As a result, already scheduled messages are prioritized over peers' locally published messages, 

```

Comment:
It's a bit ambiguous. Does "peers' locally published messages" refer to the messages published by the peer in question?

Reply:
yes, the messages published by the peer itself